internal/allocator/allocator.go

```diff
@@ -194,113 +222,116 @@ func (a *Allocator) Unassign(svc string) bool {
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, ipFamily IPFamily, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
-		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		if ipFamily != DualStack {
+			// Handle the case where the svc has already been assigned an IP but from the wrong family.
+			// This "should-not-happen" since the "ipFamily" is an immutable field in services.
+			if ipFamily != getIPFamily(alloc.ips[0]) {
+				return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ips[0].String())
+			}
 		}
-		if err := a.Assign(svc, alloc.ip, ports, sharingKey, backendKey); err != nil {
+		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
-		return alloc.ip, nil
+		return alloc.ips, nil
 	}
 
 	pool := a.pools[poolName]
 	if pool == nil {
 		return nil, fmt.Errorf("unknown pool %q", poolName)
 	}
 
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+	ips := []net.IP{}
 	for _, cidr := range pool.CIDR {
-		if cidrIsIPv6(cidr) != isIPv6 {
-			// Not the right ip-family
-			continue
-		}
-		c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
-		for pos := c.First(); pos != nil; pos = c.Next() {
-			ip := pos.IP
-			if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(ip) {
+		if len(ips) == 0 {
+			// For single stack anly address-family of cidr and clusterip needs to match
+			if ipFamily != DualStack && (getcidrIPFamily(cidr) != ipFamily) {
+				// Not the right ip-family
+				continue
+			}
+			ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svc, ports, sk)
+			if ip != nil {
+				ips = append(ips, ip)
+			}
+		} else if ipFamily == DualStack {
+			if getcidrIPFamily(cidr) == getIPFamily(ips[0]) {
 				continue
 			}
-			// Somewhat inefficiently brute-force by invoking the
-			// IP-specific allocator.
-			if err := a.Assign(svc, ip, ports, sharingKey, backendKey); err == nil {
-				return ip, nil
+			ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svc, ports, sk)
+			if ip != nil {
+				ips = append(ips, ip)
 			}
 		}
 	}
-
-	// Woops, run out of IPs :( Fail.
-	return nil, fmt.Errorf("no available IPs in pool %q", poolName)
+	if len(ips) == 0 {
```

Comment:
Here we must check also if we were asking for dual stack but we found only one good ip from the pool.